### ISSUE TITLE
gromacs_py produces gmx error alleging mismatch between gro and tpr files

### ISSUE THREAD

**Issue by hariseldon99** [samuelmurail/gromacs_py#24]
Hi all,
Installed gromacs_py from conda-forge with containerized GROMACS:2024.2 with GPU support.

## TL;DR
<a id="summary"></a>
Tied to prepare topology of a protein (zebrafish-hb) using AMBER ff in `gromacs_py`, then generate a ligand PDB (phenyl-hydrazine) using canonical SMILES info, then create ligand topology using `acpype`. Trying to add the ligand topology using the `insert_mol_sys()` to the protein `gromacs_py` object fails with the error (obtained while GROMACS ran `trjconv`) that there is a mismatch between protein GRO file and TPR file.

## TOC
1. [TL;DR](#summary) Summary of problem
2. [Setup Info](#setup) System setup
3. [Description of Sim](#desc) 
4. [Imports](#imports) Initial imports
5. [Protein](#protein) Preparation of protein
6.  [Ligand](#ligand) Preparation of Ligand
7.  [Combine](#combine) Attempt to combine the protein and ligand
8.  [Further Info](#extra) Output files

## Here is all my setup information:
<a id="setup"></a>

gromacs_py:

```python
>>>import gromacs_py
>>> gromacs_py.__version__ 
2.0.3
```
GROMACS and underlying python & OS:

```bash
$ cat /etc/issue
Ubuntu 22.04.4 LTS

$ python --version
Python 3.12.5

$ which python
/usr/local/miniforge3/envs/mdanalysis/bin/python

$ gmx --version                                                     
                   :-) GROMACS - gmx, 2024.2-conda_forge (-:

Executable:   /usr/local/miniforge3/envs/mdanalysis/bin.AVX2_256/gmx
Data prefix:  /usr/local/miniforge3/envs/mdanalysis
Working dir:  /home/$USER/.local/share/Trash/files/zeb_hb_phz_complex_sim.33
Command line:
  gmx --version

GROMACS version:     2024.2-conda_forge
Precision:           mixed
Memory model:        64 bit
MPI library:         thread_mpi
OpenMP support:      enabled (GMX_OPENMP_MAX_THREADS = 128)
GPU support:         CUDA
NBNxM GPU setup:     super-cluster 2x2x2 / cluster 8
SIMD instructions:   AVX2_256
CPU FFT library:     fftw-3.3.10-sse2-avx
GPU FFT library:     cuFFT
Multi-GPU FFT:       none
RDTSCP usage:        disabled
TNG support:         enabled
Hwloc support:       disabled
Tracing support:     disabled
C compiler:          /home/conda/feedstock_root/build_artifacts/gromacs_1720551881628/_build_env/bin/x86_64-conda-linux-gnu-cc GNU 11.4.0
C compiler flags:    -fexcess-precision=fast -funroll-all-loops -mavx2 -mfma -Wno-missing-field-initializers -O3 -DNDEBUG
C++ compiler:        /home/conda/feedstock_root/build_artifacts/gromacs_1720551881628/_build_env/bin/x86_64-conda-linux-gnu-c++ GNU 11.4.0
C++ compiler flags:  -fexcess-precision=fast -funroll-all-loops -mavx2 -mfma -Wno-missing-field-initializers -Wno-cast-function-type-strict SHELL:-fopenmp -O3 -DNDEBUG
BLAS library:        Internal
LAPACK library:      Internal
CUDA compiler:       /usr/local/cuda/bin/nvcc nvcc: NVIDIA (R) Cuda compiler driver;Copyright (c) 2005-2022 NVIDIA Corporation;Built on Wed_Sep_21_10:33:58_PDT_2022;Cuda compilation tools, release 11.8, V11.8.89;Build cuda_11.8.r11.8/compiler.31833905_0
CUDA compiler flags:-std=c++17;--generate-code=arch=compute_35,code=sm_35;--generate-code=arch=compute_37,code=sm_37;--generate-code=arch=compute_50,code=sm_50;--generate-code=arch=compute_52,code=sm_52;--generate-code=arch=compute_60,code=sm_60;--generate-code=arch=compute_61,code=sm_61;--generate-code=arch=compute_70,code=sm_70;--generate-code=arch=compute_75,code=sm_75;--generate-code=arch=compute_80,code=sm_80;--generate-code=arch=compute_86,code=sm_86;--generate-code=arch=compute_89,code=sm_89;--generate-code=arch=compute_90,code=sm_90;-Wno-deprecated-gpu-targets;--generate-code=arch=compute_53,code=sm_53;--generate-code=arch=compute_80,code=sm_80;-use_fast_math;-Xptxas;-warn-double-usage;-Xptxas;-Werror;;-Xcompiler;-fopenmp;-fexcess-precision=fast -funroll-all-loops -mavx2 -mfma -Wno-missing-field-initializers -Wno-cast-function-type-strict SHELL:-fopenmp -O3 -DNDEBUG
CUDA driver:         12.20
CUDA runtime:        11.80
```

## Description of Simulation
<a id="desc"></a>

Following [the DNA-ligand example in the docs](https://gromacs-py.readthedocs.io/en/latest/notebook/01_dna_ligand_ambertools.html), am trying to run a protein-ligand simulation using AMBER force fields.

## Step 01: Imports and setup
<a id="imports"></a>

```python
import sys
import os
import shutil

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

from gromacs_py import gmx


DATA_OUT = 'zeb_hb_phz_complex_sim'

# System Setup
vsite='none'
sys_top_folder = os.path.join(DATA_OUT, 'sys_top')
#ignore_hydrogen = {'ignh': None}

# Energy Minimisation
em_folder = os.path.join(DATA_OUT, 'em')
em_sys_folder = os.path.join(DATA_OUT, 'sys_em')
em_step_number = 10000
emtol = 10.0  	# Stop minimization when the maximum force < 10 J/mol
emstep  = 0.01      # Energy step size


# Equillibration
equi_folder = os.path.join(DATA_OUT, 'sys_equi')
HA_time = 0.5
CA_time = 1.0
CA_LOW_time = 4.0

dt_HA = 0.001
dt = 0.002

HA_step = 1000 * HA_time / dt_HA
CA_step = 1000 * CA_time / dt
CA_LOW_step = 1000 * CA_LOW_time / dt

# Production
os.makedirs(DATA_OUT, exist_ok = True)
prod_folder = os.path.join(DATA_OUT, 'sys_prod')
prod_time = 100.0

prod_step = 1000 * prod_time / dt
```

## Step 02: Protein complex
<a id="protein"></a>

Create the protein-ligand complex object with the protein PDB first. The full PDB of the protein [is here](https://github.com/hariseldon99/gromacs_sims/blob/master/zeb_hb/zeb_hb_phz_docking/input/ZEB_HB_REfined_h.pdb).

```python
pdb_file = "zeb_hb_phz_docking/input/ZEB_HB_REfined_h.pdb"
sys_name = "zeb_hb_phz_complex"
complex_sys = gmx.GmxSys(name=sys_name, coor_file=pdb_file)
complex_sys.prepare_top(out_folder=DATA_OUT, ff='amber99sb-ildn')
complex_sys.create_box(dist=1.0, box_type="dodecahedron", check_file_out=True)
complex_sys.solvate_add_ions(out_folder=DATA_OUT, name=sys_name,create_box_flag=False)
```

## Step 03: Build the ligand
<a id="ligand"></a>
### Here is my first problem, which I've ignored with `try-except1

Tried to get ligand topology using acpype. However, gromacs_py fails to import, so I do it manually.

```python
lig_name = 'phz'
lig_resname = lig_name.upper()

gmx.gmxsys.ambertools.smile_to_pdb("C1=CC=C(C=C1)NN",os.path.join(DATA_OUT,f"{lig_name}.pdb"),lig_resname)
lig_sys = gmx.GmxSys(name=lig_name, coor_file=os.path.join(DATA_OUT,f"{lig_name}.pdb"))
#acpype seems to create topology correctly, but gromacs_py does not recognize it
try:
    lig_sys.prepare_top(out_folder=os.path.join(DATA_OUT,lig_name),ff='amber99sb-ildn', include_mol={lig_resname: 'C1=CC=C(C=C1)NN'})
except:
    pass

#Since gromacs_py does not recognize the topology file, we need to set it manually
lig_sys.coor_file = os.path.join("../../",DATA_OUT,lig_name,lig_resname+".acpype",f"{lig_resname}_h_unique.pdb")
lig_sys.top_file = os.path.join("../../",DATA_OUT,lig_name,lig_resname+".acpype",f"{lig_resname}_GMX.top")
```
Output:

```bash
No amino acids present in pdb file, no PD2PQR calculation
[20:33:02] WARNING: More than one matching pattern found - picking one

acpype -i PHZ_h_unique.pdb -b PHZ -c bcc -a gaff -o gmx -n 0
gmx pdb2gmx -f 01_phz_good_his.pdb -o phz_pdb2gmx.pdb -p phz_pdb2gmx.top -i phz_posre.itp -water tip3p -ff amber99sb-ildn -ignh -vsite none
The following command could not be executed correctly :
gmx pdb2gmx -f 01_phz_good_his.pdb -o phz_pdb2gmx.pdb -p phz_pdb2gmx.top -i phz_posre.itp -water tip3p -ff amber99sb-ildn -ignh -vsite none
```

## Step 04: Combine protein and ligand
<a id="combine"></a>
### Here is my second problem:

```python
complex_sys.insert_mol_sys(lig_sys, 1, sys_name, DATA_OUT)
```

Output:

```bash
Traceback (most recent call last)
Cell In[8], [line 1](vscode-notebook-cell:?execution_count=8&line=1)
----> [1](vscode-notebook-cell:?execution_count=8&line=1) complex_sys.insert_mol_sys(lig_sys, 1, sys_name, DATA_OUT)

File /usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:3266, in GmxSys.insert_mol_sys(self, mol_gromacs, mol_num, new_name, out_folder, check_file_out)
   [3264](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:3264)     self.add_mdp(mdp_template=mini_template_mdp, mdp_options={})
   [3265](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:3265)     self.add_tpr(name="tmp")
-> [3266](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:3266) self.convert_trj(traj=False)
   [3267](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:3267) GmxSys.concat_coor(self.coor_file, mol_gromacs.coor_file,
   [3268](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:3268)                    pdb_out=concat_sys)
   [3270](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:3270) # Do the molecule insertion with the pdb_manip module:

File /usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2545, in GmxSys.convert_trj(self, name, ur, pbc, select, traj, specific_coor_out, check_file_out, tpr, **cmd_args)
   [2536](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2536) cmd_convert = os_command.Command([GMX_BIN, "trjconv",
   [2537](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2537)                                   "-f", coor_in,
   [2538](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2538)                                   "-o", coor_out,
   (...)
   [2541](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2541)                                   "-pbc", pbc],
   [2542](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2542)                                  **cmd_args)
   [2544](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2544) cmd_convert.display()
-> [2545](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2545) cmd_convert.run(com_input=select)
   [2547](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2547) if traj:
   [2548](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/gromacs_py/gmx/gmxsys.py:2548)     self.xtc = coor_out

File /usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:428, in Command.run(self, com_input, display, out_data)
    [426](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:426) print("The following command could not be executed correctly :")
    [427](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:427) self.display()
--> [428](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:428) raise RuntimeError(
    [429](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:429)     'Following Command Fails : {} \n Ret code = {} \n {} \n {}'.format(
    [430](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:430)         " ".join(self.cmd),
    [431](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:431)         proc.returncode,
    [432](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:432)         stdout_data.decode('utf-8'),
    [433](https://file+.vscode-resource.vscode-cdn.net/usr/local/miniforge3/envs/mdanalysis/lib/python3.12/site-packages/os_command_py/os_command.py:433)         stderr_data.decode('utf-8')))

RuntimeError: Following Command Fails : /usr/local/miniforge3/envs/mdanalysis/bin/gmx trjconv -f ../../zeb_hb_phz_complex_water_ion.gro -o ../../zeb_hb_phz_complex_water_ion_compact.pdb -s ../../genion_zeb_hb_phz_complex_water_ion.tpr -ur compact -pbc mol 
 Ret code = 1 
 Note that major changes are planned in future for trjconv, to improve usability and utility.
Select group for output
Selected 0: 'System'
 
                :-) GROMACS - gmx trjconv, 2024.2-conda_forge (-:

Executable:   /usr/local/miniforge3/envs/mdanalysis/bin.AVX2_256/gmx
Data prefix:  /usr/local/miniforge3/envs/mdanalysis
Working dir:  /home/$USER/gitrepos/gromacs_sims/zeb_hb/zeb_hb_phz_complex_sim/phz/zeb_hb_phz_complex_sim
Command line:
  gmx trjconv -f ../../zeb_hb_phz_complex_water_ion.gro -o ../../zeb_hb_phz_complex_water_ion_compact.pdb -s ../../genion_zeb_hb_phz_complex_water_ion.tpr -ur compact -pbc mol

Will write pdb: Protein data bank file
Reading file ../../genion_zeb_hb_phz_complex_water_ion.tpr, VERSION 2024.2-conda_forge (single precision)
Reading file ../../genion_zeb_hb_phz_complex_water_ion.tpr, VERSION 2024.2-conda_forge (single precision)
Group     0 (         System) has 21639 elements
Group     1 (        Protein) has  2220 elements
Group     2 (      Protein-H) has  1095 elements
Group     3 (        C-alpha) has   143 elements
Group     4 (       Backbone) has   429 elements
Group     5 (      MainChain) has   573 elements
Group     6 (   MainChain+Cb) has   708 elements
Group     7 (    MainChain+H) has   711 elements
Group     8 (      SideChain) has  1509 elements
Group     9 (    SideChain-H) has   522 elements
Group    10 (    Prot-Masses) has  2220 elements
Group    11 (    non-Protein) has 19419 elements
Group    12 (          Water) has 19419 elements
Group    13 (            SOL) has 19419 elements
Group    14 (      non-Water) has  2220 elements
Select a group: Reading frames from gro file 'Protein in water', 21569 atoms.

Reading frame       0 time    0.000   
Precision of ../../zeb_hb_phz_complex_water_ion.gro is 0.001 (nm)

-------------------------------------------------------
Program:     gmx trjconv, version 2024.2-conda_forge
Source file: src/gromacs/tools/trjconv.cpp (line 1037)

Fatal error:
Index[21569] 21570 is larger than the number of atoms in the
trajectory file (21569). There is a mismatch in the contents
of your -f, -s and/or -n files.

For more information and tips for troubleshooting, please check the GROMACS
website at https://manual.gromacs.org/current/user-guide/run-time-errors.html
-------------------------------------------------------
```

### Further Information:
<a id="extra"></a>

1. The final GRO file is [here](https://github.com/hariseldon99/gromacs_sims/blob/master/zeb_hb/zeb_hb_phz_complex_sim/zeb_hb_phz_complex_water_ion.gro)

2. The final TPR file is [here](https://github.com/hariseldon99/gromacs_sims/blob/master/zeb_hb/zeb_hb_phz_complex_sim/genion_zeb_hb_phz_complex_water_ion.tpr)

4. The output of `gmx dump` on the tpr file is [here](https://mega.nz/file/dkwnESiL#P5wVtW0oQN3Rw49keLT8cnzdDb7TpBAmix2avEkbzWY)

9. Ligand stuff generated by `acpype` is [here](https://github.com/hariseldon99/gromacs_sims/tree/master/zeb_hb/zeb_hb_phz_complex_sim/phz).

Near as I can see, the GRO file contains the proper number of atoms, which is 21569, but I cannot make sense of the text dump of the TPR. Can someone explain to me what's going wrong?
